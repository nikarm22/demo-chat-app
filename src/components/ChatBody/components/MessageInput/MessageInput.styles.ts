import { makeStyles } from "@material-ui/core";

export const useStyles = makeStyles(() => ({
    root: {
        display: 'flex',
    },
    input: {
        flex: 1,
    },
}));
