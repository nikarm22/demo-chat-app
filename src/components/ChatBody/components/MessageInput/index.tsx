import { FormControl, Input, InputAdornment, TextField } from '@material-ui/core';
import { IconButton, makeStyles, Paper } from '@material-ui/core';
import SendIcon from '@material-ui/icons/Send';
import React from 'react';

interface IMessageInputProps {
    onSend?: (message: string) => void;
}

const useStyles = makeStyles(() => ({
    root: {
        display: 'flex',
    },
    input: {
        flex: 1,
    },
}));

const MessageInput: React.FC<IMessageInputProps> = props => {
    const classes = useStyles();

    return (
        <FormControl component="form" className={classes.root} variant="filled">
            <Input
                className={classes.input}
                autoFocus
                endAdornment={
                    <InputAdornment position="end">
                        <IconButton
                            type="submit"
                        >
                            <SendIcon />
                        </IconButton>
                    </InputAdornment>
                }
            />
        </FormControl>
    );
};

export { MessageInput };
