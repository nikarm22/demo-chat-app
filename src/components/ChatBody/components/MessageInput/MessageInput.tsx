import { FormControl, Input, InputAdornment } from '@material-ui/core';
import IconButton from '@material-ui/core/IconButton';
import SendIcon from '@material-ui/icons/Send';
import React, { useCallback, useState } from 'react';
import { useStyles } from './MessageInput.styles';

interface IMessageInputProps {
    onSend?: (message: string) => void;
}

const MessageInput: React.FC<IMessageInputProps> = props => {
    const { onSend } = props;
    const [currentMessage, setCurrentMessage] = useState('');
    const classes = useStyles();

    const handleChange = useCallback((event) => {
        setCurrentMessage(event.target.value);
    }, []);

    const handleSubmit = useCallback((event) => {
        event.preventDefault();
        if(currentMessage.trim().length) {
            onSend?.(currentMessage.trim());
            setCurrentMessage('');
        }
    }, [currentMessage, onSend]);

    return (
        <FormControl component="form" className={classes.root} variant="filled" onSubmit={handleSubmit}>
            <Input
                name="message"
                className={classes.input}
                autoFocus
                value={currentMessage}
                onChange={handleChange}
                data-testid="message-input"
                endAdornment={
                    <InputAdornment position="end">
                        <IconButton
                            data-testid="send-message"
                            type="submit"
                        >
                            <SendIcon />
                        </IconButton>
                    </InputAdornment>
                }
            />
        </FormControl>
    );
};

export { MessageInput };
