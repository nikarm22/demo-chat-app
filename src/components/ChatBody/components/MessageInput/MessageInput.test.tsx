import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { MessageInput } from './MessageInput';

describe('<MessageInput />', () => {
    it('should match the snapshot', async () => {
        const { asFragment } = render(<MessageInput />);

        expect(asFragment()).toMatchSnapshot();
    });

    it('should call onSend method when clicked on send buttons and when input field contains text', async () => {
        const mockOnSend = jest.fn();

        const { findByTestId } = render(
            <MessageInput onSend={mockOnSend} />
        );

        const button = await findByTestId('send-message');
        const input =  (await findByTestId('message-input')).querySelector('input');

        const mockValue = 'test';

        if (input) fireEvent.change(input, { target: { value: mockValue }});
        fireEvent.click(button);

        expect(mockOnSend).toBeCalledTimes(1);
        expect(mockOnSend).toBeCalledWith(mockValue);
    });

    it('should not call onSend method when the message text is empty', async () => {
        const mockOnSend = jest.fn();

        const { findByTestId } = render(
            <MessageInput onSend={mockOnSend} />
        );

        const button = await findByTestId('send-message');
        fireEvent.click(button);

        expect(mockOnSend).toBeCalledTimes(0);
    });

    it('should not call onSend method when the message text contains only spaces', async () => {
        const mockOnSend = jest.fn();

        const { findByTestId } = render(
            <MessageInput onSend={mockOnSend} />
        );

        const button = await findByTestId('send-message');
        const input =  (await findByTestId('message-input')).querySelector('input');

        const mockValue = '           ';

        if (input) fireEvent.change(input, { target: { value: mockValue }});
        fireEvent.click(button);

        expect(mockOnSend).toBeCalledTimes(0);
    });

    it('should call the onSend method when Enter is pressed', async () => {
        const mockOnSend = jest.fn();

        const { findByTestId } = render(
            <MessageInput onSend={mockOnSend} />
        );

        const input =  (await findByTestId('message-input')).querySelector('input');

        const mockValue = 'test';

        if (input) {
            fireEvent.change(input, { target: { value: mockValue }});
            fireEvent.submit(input);
        }

        expect(mockOnSend).toBeCalledTimes(1);
        expect(mockOnSend).toBeCalledWith(mockValue);
    });
});
