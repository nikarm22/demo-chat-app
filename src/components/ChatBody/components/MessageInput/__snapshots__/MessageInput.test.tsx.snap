// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageInput /> should match the snapshot 1`] = `
<DocumentFragment>
  <form
    class="MuiFormControl-root makeStyles-root-1"
  >
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline makeStyles-input-2 Mui-focused Mui-focused MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
      data-testid="message-input"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
        name="message"
        type="text"
        value=""
      />
      <div
        class="MuiInputAdornment-root MuiInputAdornment-filled MuiInputAdornment-positionEnd"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root"
          data-testid="send-message"
          tabindex="0"
          type="submit"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M2.01 21L23 12 2.01 3 2 10l15 2-15 2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
